src/java/com/puppetlabs/jruby_utils/pool/JRubyPool.java

```diff
@@ -399,8 +427,8 @@ private void signalIfAllRegisteredInQueue() {
         // the pool lock but is waiting for all registered elements to be
         // returned to the queue.
         if (registeredElements.size() == liveQueue.size() ||
-                poisonPill != null) {
-            allRegisteredInQueue.signal();
+                pill != null) {
+            lockAvailable.signal();
```

Comment:
Could also change the name of this method to something more inclusive, like `signalIfAllRegisteredOrPillInQueue` or `signalIfLockCanProceed`.  I'm good with whatever seems most reasonable to you here.